core/src/main/java/org/infinispan/commands/topology/ConsistentHashUpdateCommand.java

```diff
@@ -0,0 +1,138 @@
+import org.infinispan.distribution.ch.ConsistentHash;
+import org.infinispan.factories.GlobalComponentRegistry;
+import org.infinispan.partitionhandling.AvailabilityMode;
+import org.infinispan.remoting.responses.SuccessfulResponse;
+import org.infinispan.remoting.transport.Address;
+import org.infinispan.topology.CacheTopology;
+import org.infinispan.topology.PersistentUUID;
+
+/**
+ * Coordinator to member:
+ * The coordinator is updating the consistent hash.
+ * Used to signal the end of rebalancing as well.
+ *
+ * @author Ryan Emerson
+ * @since 11.0
+ */
+public class ConsistentHashUpdateCommand extends AbstractTopologyCommand {
+
+   public static final byte COMMAND_ID = 95;
+
+   private String cacheName;
+   private Address sender;
+   private ConsistentHash currentCH;
+   private ConsistentHash pendingCH;
+   private CacheTopology.Phase phase;
+   private List<Address> actualMembers;
+   private List<PersistentUUID> persistentUUIDs;
+   private AvailabilityMode availabilityMode;
+   private int rebalanceId;
+   private int topologyId;
+   private int viewId;
+
+   // For CommandIdUniquenessTest only
+   public ConsistentHashUpdateCommand() {
+      super(COMMAND_ID);
+   }
+
+   public ConsistentHashUpdateCommand(String cacheName, Address sender, CacheTopology cacheTopology,
+                                      AvailabilityMode availabilityMode, int viewId) {
+      super(COMMAND_ID);
+      this.cacheName = cacheName;
+      this.sender = sender;
+      this.topologyId = cacheTopology.getTopologyId();
+      this.rebalanceId = cacheTopology.getRebalanceId();
+      this.currentCH = cacheTopology.getCurrentCH();
+      this.pendingCH = cacheTopology.getPendingCH();
+      this.phase = cacheTopology.getPhase();
+      this.availabilityMode = availabilityMode;
+      this.actualMembers = cacheTopology.getActualMembers();
+      this.persistentUUIDs = cacheTopology.getMembersPersistentUUIDs();
+      this.viewId = viewId;
+   }
+
+   @Override
+   public CompletionStage<?> invokeAsync(GlobalComponentRegistry gcr) throws Throwable {
+      CacheTopology topology = new CacheTopology(topologyId, rebalanceId, currentCH, pendingCH, phase, actualMembers, persistentUUIDs);
+      return gcr.getLocalTopologyManager()
+            .handleTopologyUpdate(cacheName, topology, availabilityMode, viewId, sender)
+            .thenApply(SuccessfulResponse::create);
+   }
+
+   public String getCacheName() {
+      return cacheName;
+   }
+
+   public ConsistentHash getCurrentCH() {
+      return currentCH;
+   }
+
+   public ConsistentHash getPendingCH() {
+      return pendingCH;
+   }
+
+   public CacheTopology.Phase getPhase() {
+      return phase;
+   }
+
+   public int getTopologyId() {
+      return topologyId;
+   }
+
+   @Override
+   public void writeTo(ObjectOutput output) throws IOException {
+      MarshallUtil.marshallString(cacheName, output);
+      output.writeObject(sender);
+      output.writeObject(currentCH);
```

Comment:
`CacheTopology` has an externalizer. You can extract the code to static methods and invoke them. No need to declare all these fields neither copy-paste the write/read code.

Reply:
The `CacheTopology` externalizer will be removed, as well this `writeTo` implementation once the GlobalMarshaller is converted to protostream. I just lifted this code from the old command.

Reply:
I noticed the "copy-paste" style :) 
How long until it is converted to protostream? Is it worth to make the change/refactorization?

Reply:
I'm refactoring the commands to make the protostream conversion easier, so it's actually already started :slightly_smiling_face: